### ISSUE TITLE
State of readiness of the cases for the paper

### ISSUE THREAD

**Issue by philipcardiff** [philipcardiff/paper-newton-krylov#3]
| Case    | Ready | Meshes | Comments  |
|---------|-----|-----|--------|
| MMS   | [x]  | Scripts ready; Structured hex (blockMesh) or tet/poly (Gmsh + polyDualMesh) | 2nd order displacement; 1-1.5 order stresses. I have not checked unstructured tet/poly  |
| Spherical cavity | [ ]  | Scripts ready; poly mesh (Gmsh + polyDualMesh) | runs but mesh convergence needs check; this case may be superseded by the MMS case so we can leave this out (a pity given all my effort on the analytical solution!)     |
| Elliptic plate | [x] | Scripts ready. Hex mesh (blockMesh). | Case is ready. |
| narrowTmember | [x] | Scripts ready. Hex mesh (Gmsh) | Case is ready. |
| Ventricle | [x] | Scripts ready. Structured hex/prism mesh (blockMesh + extrudeMesh) | Case runs with SNES; verification of mesh convergence vs reference not yet checked |
| Cook's membrane (linear elastic)     | []  | Scripts ready; Structured hex (blockMesh) | Runs and gives reasonable answers (the scale seems off for this benchmark ... very large displacements... maybe E should be x1000)    |
| Cook's membrane (hyperelastic)     | []  | Scripts ready; Structured hex (blockMesh) | Checked with Neo-Hookean. The prediction seems OK for neo-Hooekan   |
| Cook's membrane (hyperelastoplastic)     | [ ]  | Scripts ready; Structured hex (blockMesh) | Case is set up and runs but crashes near the end for finer meshes, as the entire domain becomes plastic (I suspect our precondition matrix is not very good in this limit)  |
| Cantilever vibration | [x] | Hex mesh (blockMesh) | High load to cause significant geometric nonlinearity |
| Turbine blade vibration | [ ] | Fluent poly mesh from Zeljko | Complex mesh not running but less complex does run |
| Hron-Turek FSI | [ ] | three meshes ready | I am waiting on the finest mesh to run |

Other comments:
- Above, we have the following physics:
  - Static and dynamic
  - Linear geometry and nonlinear geometry
  - Linear elasticity and hyperelasticity
  - Various mesh types
  - Various geometry types, from simple to more complex
- But we do not have elastoplasticity, which would be easy to include and nice to show: we could run Cook's membrane as elastoplastic too (@iBatistic thoughts?).

**Comment by iBatistic** [comment]
> But we do not have elastoplasticity, which would be easy to include and nice to show: we could run Cook's membrane as elastoplastic too (@iBatistic thoughts?).

Why not, I will also add cases for linear-elastic material. Later, we simply do not include what we consider redundant.

I opened a new branch called `ivan`, when things accumulate, I'll open a PR. The things I added are:
- Goodier analytical solution for displacement
- Mesh and geometry figures for Cook membrane. Geomery image for ventricle case (maybe it will be useful when refering in postprocessing diagrams, anyway we can remove if it is not useful)
- Missing references for Ventricle case

**Comment by iBatistic** [comment]
> Runs and gives reasonable answers (the scale seems off for this benchmark ... very large displacements... maybe E should be x1000)

The problem is that the domain depth is set to 1 mm instead of 1 m. To correct this, we need to scale the traction to 6.25.

**Comment by philipcardiff** [comment]
Ah, OK.

But, since a traction is used (force per unit area), the solution is not a function of the depth (as the total force increases as the depth increases). If we instead used a force then scaling the depth would work (assuming I am not misunderstanding). Nonetheless, using 6.25 Pa for the traction will work and the problem is linear so the answers just scale linearly anyway.

**Comment by iBatistic** [comment]
You are totally right. We just need to set domain depth to be one and the solution is fine (checked). I'm currently working with Cooks membrane, later today I will have PR which will correct this. 

**EDIT:**
I wrote above that changing depth and leaving 6250Pa for load will get right solution; which is wrong as you explained.

The correct parameters should be a Young's modulus of 70 MPa and a load of 6.25 MPa. CoFEA's page has an incorrect case description, but I found the correct input files on their GitHub page, where the values match those mentioned above.

**Comment by iBatistic** [comment]
Philip, I noticed that when I run the cases, more than one core is being utilized. I suspect this is due to PETSc's internal multithreading.

On my installation, the `solids4Foam` command uses all available cores, while `mpirun -np 1` runs on two cores and is slightly faster. **EDIT**: tried some one another case  and `solids4Foam` is using 3 cores

Initially, I thought this might explain why the PETSc segregated solver is faster than the OpenFOAM solver. However, the PETSc segregated solver appears to be using only one core. 

I also tested using `export OMP_NUM_THREADS=1`, and with this, PETSc uses just one core.
I tried with `OMP_NUM_THREADS=1` on one case, and the clock time was the same as when multiple cores were being used.

Do you observe the same behavior with your installation?


![Screenshot from 2024-09-18 09-36-18](https://github.com/user-attachments/assets/06f4180e-de5f-45b9-b5d6-b79a2198bc0f)

**UPDATE**

I tested on my workstation, but I didn't encounter the described behavior. It appears to be related specifically to my laptop installation. Anyhow, I wanted to inform you so you can check this for your system as well.

**Comment by philipcardiff** [comment]
Hi Ivan,

Yes, I have noticed this behaviour on some of my installations. I looked into it, and it depends on how PETSc is configured. If PETSc is configured with OpenMP then it seems to use all the available cores for at least some of the linear solvers, e.g. the MUMPS/LU direct solvers use all the cores.

But good point: we should be careful when reporting and comparing times that we are using only one core or the same number of cores.

**Comment by iBatistic** [comment]
@philipcardiff, please let me know if there's anything specific I can help with in the paper at the moment.

**Comment by philipcardiff** [comment]
Thanks @iBatistic.

Section 4.1 is almost ready:
- *Axial turbine blade*: I will try another mesh for the complex blade; if it doesn't work, I will revert to the simpler blade
- *Hron-Turek*: the results seem to be good enough; I am re-running it but a tighter FSI tolerance but this will not change the results much
- *Cook's membrane*: 
  - Linear elastic: good agreement
  - Hyperelastic and hyperelastoplastic: our results appear to be converging to a different answer than the reference results. Do you know why? Reviewers may question if our method is correct in these cases. I will run the cases in Abaqus and see what results it gives. The good thing is that these are quick cases.

I am now starting to look at the remainder of the paper (Sections 4.2-4.7). We can probably use a combination of tables and figures to show the speed comparisons. Feel free to make suggestions. Once I start filling it out more, I may ask for help.

**Comment by iBatistic** [comment]
> Hyperelastic and hyperelastoplastic: our results appear to be converging to a different answer than the reference results. Do you know why? Reviewers may question if our method is correct in these cases. I will run the cases in Abaqus and see what results it gives. The good thing is that these are quick cases.

I'm not sure why we have this discrepancy, but it's not related to snes. As you said, it could potentially raise concerns with reviewers. Let's wait for the results from Abaqus, and based on that, we can decide whether to include this case in the paper. 

> I am now starting to look at the remainder of the paper (Sections 4.2-4.7). We can probably use a combination of tables and figures to show the speed comparisons. Feel free to make suggestions. Once I start filling it out more, I may ask for help.

Sure, ok

**Comment by philipcardiff** [comment]
I think I found one reason for the difference: at [deal.ii](https://www.dealii.org/current/doxygen/deal.II/code_gallery_Quasi_static_Finite_strain_Compressible_Elasticity.html#:~:text=Overview,ansatz%20under%20quasi%2Dincompressible%20conditions), I noticed they say *and a uniform traction load is applied at the other end such that the total force acting on this surface totals 1 Newton.* This means the traction is applied per unit **un**deformed area, where `solidTraction` applies the traction per deformed area. This probably explains why our case predicts a different displacement than the deal.ii case.

Abaqus using ~1 mm quadratic elements gives:
- 14.310 mm using the deformed area: same as `solidTraction` approach. I need to do a mesh study to see if it converges to the same value as we are getting.
- 13.7725 mm using the undeformed area: I thought this would be closer to deal.ii results but it is further away so I am a bit confused...

OK, I looked further into [deal.ii page](https://www.dealii.org/current/doxygen/deal.II/code_gallery_Quasi_static_Finite_strain_Compressible_Elasticity.html#:~:text=Overview,ansatz%20under%20quasi%2Dincompressible%20conditions) (... I am slowly writing this post as I watch TV...); in particular, I looked at how they calculate the stress for a neo-Hookean material. They calculate the Kirchhoff volumetric stress as
```
      SymmetricTensor<2,dim,NumberType>
      get_tau_vol(const NumberType &det_F) const
      {
        return NumberType(get_dPsi_vol_dJ(det_F) * det_F) * Physics::Elasticity::StandardTensors<dim>::I;
      }
```
where
```
      NumberType
      get_dPsi_vol_dJ(const NumberType &det_F) const
      {
        return (kappa / 2.0) * (det_F - 1.0 / det_F);
      }
```
so the volumetric Kirchoff stress is (subbing 2 into 1 and replacing `det_F` with `J`)
```
    (kappa / 2.0) * (J - 1.0 / J)* J
==  (kappa / 2.0) * (J - 1.0)
==  0.5*K*(J - 1.0)
```

But we calculate it as
```
0.5*K()*(pow(J, 2.0) - 1.0)
```

So, they are using a different definition of pressure than us. I think this may also be a reason for the differences we are seeing.

By definition, "neo-Hookean" only defines the deviatoric response so they (and we) are not wrong in how we define the pressure/volumetric response, as we are free to choose whatever form we like (as long as it reduces to the small strain version in the limit). This is normally not an issue as neo-Hookean is often used in the incompressible or quasi-incompressible form, where $J \approx 1$ so the volumetric strains are small and all the definitions coincide.

So.... what do we do... maybe the easier is for us to:
- add the alternative pressure definition to neoHookeanElastic
- add a "useUndeformedArea" to solidTraction/nonLinGeomTL/UL
and see if we get a better agreement...

**Comment by philipcardiff** [comment]
I changed the pressure definition (but not solidTraction) and got:
```
30 -0.016383 0.0156883 0 0.0226832  
```
where our original approach gave
```
30 -0.0141313 0.0141904 0 0.0200265
```
so it definitely makes a difference.

We are now overshooting the deal.ii value of `14.74`.

Hopefully, if I also use the undeformed areas in solidTraction it will bring us close...

**Comment by philipcardiff** [comment]
OK, with the alternate pressure approach and solidTraction using the undeformed areas I get
```
30 -0.0156737 0.0151653 0 0.0218094
```
for mesh 4.

We are closer, but there is still a difference ... hmm. It is interesting that Abaqus also gives a different answer, but they seem to use another [definition of pressure](https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/stm/default.htm?startat=ch04s06ath123.html).

I will push the changes to the code and case. Feel free to have a look!

**Comment by philipcardiff** [comment]
Looking back, I see I made a mistake in the second line of
```
    (kappa / 2.0) * (J - 1.0 / J)* J
==  (kappa / 2.0) * (J - 1.0)
==  0.5*K*(J - 1.0)
```
It should be 
```
    (kappa / 2.0) * (J - 1.0 / J)* J
==  (kappa / 2.0) * ((J^2 - 1.0)/J)*J
==  0.5*kappa * (J^2 - 1.0)
```
which is actually the same definition we use for the Kirchhoff volumetric stress!
In any case, it would need to be `kappa * (J - 1.0)` (without the 0.5) to be consistent in the small strain limit.

If we only use the change in solidTraction, we get
```
30 -0.0136162 0.0137893 0 0.019379
```
which is further away from the deal.ii reference....

**Comment by philipcardiff** [comment]
But actually, Abaqus gives `13.7725` mm using the undeformed area (for the single quadratic element I tried), which is very close!

I suspect we are misinterpreting something in the deal.ii case or material definition, but if Abaqus agrees with the mesh refinement limit, then we can use that.

**Comment by philipcardiff** [comment]
I was about to go to bed but then I found the following text on the deall.ii site:
>  Specifically, figure 5 of Miehe (1994) plots the vertical displacement at the midway point on the traction surface for the compressible 3d case.

The displacement is measured at the midway point on the traction surface, not the top corner.

Confusingly, the deall.ii page gives a table titled *Tip y-displacement (in mm)* but I think this is actually the midway point y-displacement.

I ran our model (using the undeformed area option) and it gives the following for the midway point:
```
30 -0.00917811 0.0147262 0 0.0173522
```
which is very close to the converged value given by deal.ii: 14.74 mm.

I *think* it now makes sense.

Time to sleep.

**Comment by iBatistic** [comment]
@philipcardiff, it was interesting going through these posts! 😄 I'm glad you managed to solve it. I apologize for not carefully reading the text earlier, which caused some unnecessary confusion and wasted your time. Typically, the literature provides results for the tip, so it's unclear to me why they gave it for the midpoint in this case.

**Comment by philipcardiff** [comment]
Ha,  no problem. It was an interesting journey!

I suspect the elastoplastic case has a similar journey, although hopefully shorter.

**Comment by iBatistic** [comment]
@philip in the latest commit (#fda35b2) in the Cook's membrane PR (benchmark repository), I have added a script to check the effects of the Rhie-Chow scaling factor. The `AllrunMeshStudy` script will run simulations with different scaling factor values on the specified mesh, while `AllpostScalingStudy` will extract time and memory data.
Inputs for the scripts are defined in `USER_VARIABLE_NAMES`.

The scripts are currently located in the `narrowTmember` case, but they can be used with any case to generate data for Table 6. If you'd like, let me know which case you want to use to assess the effect on accuracy, and I can modify or create a script for it accordingly.

**Comment by philipcardiff** [comment]
@iBatistic, the paper is ready for your review (when you are back to work in the new year)! I will also send it to Željko and Hiro (who may be interested in being a co-author - I have added him for now). I will also ask Dylan if he is interested in adding Abaqus results, e.g. for the cantilever case, so we have a reference solution.

I expect this can be submitted by mid-January, depending on the comments.

**Comment by iBatistic** [comment]
Hi @philipcardiff, today I'm back at work! :) Reviewing this paper will be a great way to start the new year. I’ll post here my comments by tomorrow.

P.S. Wishing you all the best in the new year!

**Comment by iBatistic** [comment]
Hi Philip, I have gone through the paper, and it is impressive and beautiful. I only noticed a few minor things:

- The full name of the abbreviation GMRES is not mentioned in the text
- It would be nice to move the keywords to the first page, if possible.
- In equations (2), (3), and (4), the differential symbol  𝑑Γ dΓ has an italicized 'd'. I prefer it upright to slightly increase the distance between 'd' and 'Γ', but you can leave it as is.
- The same as above, I think that Transpose symbol T is more ofter writen upright than italic (equation for _F_ and _f_)
- equation 8: should $u_k$ be instead of $u_n$. 
- It seems to me that the letters in Figure 1 are a bit small compared to the rest of the text. The subscript "i" is therefore a little harder to see
- In this sentece: "difference schemes, e.g. first-order Euler, second-order backwards, second-order Newmark-beta"
should "or" be added before the last one.
- In sentece "square braces on thr right-hand side of Equation 28,"  thr -> the
- Perhaps you could consider using \dfrac instead of \frac in equations 18 and 38. The scaling of the symbols looks a bit off to me
- For case 5, you wrote the density as  kg m−3, while in other cases, it is expressed as  kg/m^3 
- "the unified memory architecture of the M1 Ultra CPU likely provides greater memory bandwidth" I think memory speed is a bigger factor than CPU speed. The M1 chip has a memory speed of 6400, while the AMD chip (which I assume uses DDR4) has a memory speed around 3000. Based on this, the M1 chip should be roughly twice as fast for simulations
- This text was not clear to me: "Figure 19(b), the time required for 1 core has been estimated for the segregated and Jacobian-free Newton-Krylov LU preconditioner approaches as 8× the time required for 8 cores; consequently, the absolute values of speedup should be ignored (or carefully interpretted) and instead the trend is of greater value."
- Just looking in Appendix B, everything look fine to me,  I’ll try deriving the expressions by hand, but frankly, I expect to arrive at something different from your results. Perhaps Željko will have a better understanding of why previous discussions authors refered to it as third order


References
- The name Željko is missing in references [9] and [11]
- In reference [38], it should be Karač instead of Karac.
- In references [59], [61], and [47], there are '???' marks before the year.
- In the title of reference [11], 'OpenFOAM' should be capitalized instead of 'openfoam'.
- In references [19], [20], [21], [22], [23], and [30], the names 'Newton,' 'Krylov,' and 'Stokes' should start with capital letters

**Comment by philipcardiff** [comment]
Great; thank you! I will incorporate these changes and those from the others. Hopefully, I can submit it in the coming couple of weeks.

**Comment by philipcardiff** [comment]
Checklist

* [x] The full name of the abbreviation GMRES is not mentioned in the text: *defined in the abstract and in the main body after first use*
* [x] It would be nice to move the keywords to the first page, if possible. *this template does not let me control this, but in any case Hiro would rather be acknowledged than a co-author so this fixes it*
* [ ] In equations (2), (3), and (4), the differential symbol  𝑑Γ dΓ has an italicized 'd'. I prefer it upright to slightly increase the distance between 'd' and 'Γ', but you can leave it as is. *I got tired of adding `\text{d}` and I noticed many other authors use the italicised version so I will leave it; I suspect the journal will set it based on their own style guides anyway* 
* [x] The same as above, I think that Transpose symbol T is more ofter writen upright than italic (equation for _F_ and _f_)
* [x] equation 8: should u_k  be instead of u_n
* [x] It seems to me that the letters in Figure 1 are a bit small compared to the rest of the text. The subscript "i" is therefore a little harder to see
* [x] In this sentece: "difference schemes, e.g. first-order Euler, second-order backwards, second-order Newmark-beta"; should "or" be added before the last one.
* [x] In sentece "square braces on thr right-hand side of Equation 28,"  thr -> the
* [x] Perhaps you could consider using \dfrac instead of \frac in equations 18 and 38. The scaling of the symbols looks a bit off to me
* [x] For case 5, you wrote the density as  kg m−3, while in other cases, it is expressed as  kg/m^3
* [x] "the unified memory architecture of the M1 Ultra CPU likely provides greater memory bandwidth" I think memory speed is a bigger factor than CPU speed. The M1 chip has a memory speed of 6400, while the AMD chip (which I assume uses DDR4) has a memory speed around 3000. Based on this, the M1 chip should be roughly twice as fast for simulations. *Good point. According to https://en.wikipedia.org/wiki/Apple_M1#:~:text=The%20M1%20Ultra%20doubles%20the,GB%20or%20128%20GB%20configuration. the memory bandwidth of M1 Ultra is actually 800 GB/s. I have updated the text to be 'The difference in performance can be primarily attributed to the unified memory architecture of the M1 Ultra, which provides significantly greater memory bandwidth compared to the DDR4 memory used in the MeluXina compute nodes.
'. I have decided not to include the specific numbers as it is hard to know what the specific memory bandwidth available to a single core of each CPU*
* [x] This text was not clear to me: "Figure 19(b), the time required for 1 core has been estimated for the segregated and Jacobian-free Newton-Krylov LU preconditioner approaches as 8× the time required for 8 cores; consequently, the absolute values of speedup should be ignored (or carefully interpretted) and instead the trend is of greater value.". *Here I did not run the segregated and LU cases in serial so I don't know what the 1 core time is. Hence I cannot calculate the speed-up. To overcome this, I estimated the 1 core time as 8x the 8 core time. However, I agree this is confusing and this sentence is strange; I will remove the segregated and LU results from the speed-up graph*
* [x] Just looking in Appendix B, everything look fine to me,  I’ll try deriving the expressions by hand, but frankly, I expect to arrive at something different from your results. Perhaps Željko will have a better understanding of why previous discussions authors refered to it as third order

> References
* [x] The name Željko is missing in references [9] and [11]
* [x] In reference [38], it should be Karač instead of Karac.
* [x] In references [59], [61], and [47], there are '???' marks before the year. *this was caused by a missing `address` field, even though there was no warning*
* [x] In the title of reference [11], 'OpenFOAM' should be capitalized instead of 'openfoam'.
* [x] In references [19], [20], [21], [22], [23], and [30], the names 'Newton,' 'Krylov,' and 'Stokes' should start with capital letters

**Comment by philipcardiff** [comment]
Resolved in 4f95d11
